lms/djangoapps/certificates/api.py

```diff
@@ -463,51 +463,66 @@ def get_active_web_certificate(course, is_preview_mode=None):
-    if course_organization:
-        org_id = course_organization[0]['id']
+    org_id = get_course_organization_id(course_key)
 
-    if org_id and mode:
-        template = CertificateTemplate.objects.filter(
+    # only consider active templates
+    active_templates = CertificateTemplate.objects.filter(is_active=True)
+
+    if org_id and mode:  # get template by org, mode, and key
+        org_mode_and_key_templates = active_templates.filter(
             organization_id=org_id,
-            course_key=course_key,
             mode=mode,
-            is_active=True
+            course_key=course_key
         )
-    # if don't template find by org and mode
-    if not template and org_id and mode:
-        template = CertificateTemplate.objects.filter(
+        template = get_templates_with_language_or_defaults(language, org_mode_and_key_templates)
+
+    # since no template matched that course_key, only consider templates with empty course_key
+    empty_course_key_templates = active_templates.filter(course_key=CourseKeyField.Empty)
+    if not template and org_id and mode:  # get template by org and mode
+        org_and_mode_templates = empty_course_key_templates.filter(
             organization_id=org_id,
-            course_key=CourseKeyField.Empty,
-            mode=mode,
-            is_active=True
+            mode=mode
         )
-    # if don't template find by only org
-    if not template and org_id:
-        template = CertificateTemplate.objects.filter(
+        template = get_templates_with_language_or_defaults(language, org_and_mode_templates)
+    if not template and org_id:  # get template by only org
+        org_templates = empty_course_key_templates.filter(
             organization_id=org_id,
-            course_key=CourseKeyField.Empty,
-            mode=None,
-            is_active=True
+            mode=None
         )
-    # if we still don't template find by only course mode
-    if not template and mode:
-        template = CertificateTemplate.objects.filter(
+        template = get_templates_with_language_or_defaults(language, org_templates)
+    if not template and mode:  # get template by only mode
+        mode_templates = empty_course_key_templates.filter(
             organization_id=None,
-            course_key=CourseKeyField.Empty,
-            mode=mode,
-            is_active=True
+            mode=mode
         )
-
+        template = get_templates_with_language_or_defaults(language, mode_templates)
     return template[0].template if template else None
 
 
+def get_templates_with_language_or_defaults(language, templates):
+    """
+    Returns templates that match passed in language.
+    Returns default templates If no language matches, or language passed is None
+    """
+    two_letter_language = _get_two_letter_language_code(language)
+    language_templates = templates.filter(language=two_letter_language)
+    return language_templates if language_templates else templates.filter(language=None)
```

Comment:
Rather than checking for the two-letter language first, and then falling back to a second query to check for null language, could we use `WHERE ... OR ...` to check both at the same time?

Reply:
I think you can use Django Q objects to do that. https://docs.djangoproject.com/en/1.11/topics/db/queries/#complex-lookups-with-q-objects

Reply:
I dont think this is exactly a normal OR check here, we don't want the set of all templates with a null language OR a specific language, we want the ONLY the templates with a specific language, if none of those exist then we want the 'default' templates with null language

Reply:
I think it's still probably more efficient to get two templates back in that case and return the one with a non-null language if present, than it would be to run two queries.

Reply:
I'm not sure how to do this with 1 query. In order to check if a language_specific template exists there will need to be a DB hit. Even using an OR I think its the same number of hits. Below is what I think you are suggesting, but I don't think its more efficient. 

language_templates = templates.filter(Q(lang=two_letter_lang) | Q(lang=null))
if language_templates.filter(language=two_letter_language) 
    return language_templates.filter(language=two_letter_lang)
else:
    return language_templates.filter(language=null)

Were you thinking of something else?

Reply:
This way will still result in 2 DB hits.  The way to do it would be to actually resolve the results from your first line and filter them in python rather than in an another query

```
language_templates = templates.filter(Q(lang=two_letter_lang) | Q(lang=null)))
result_with_lang = [template for template in language_templates if template.lang not null]
if result_with_lang:
    return result_with_lang
else:
    return [template for template in language_templates if template.lang == null]
```

Reply:
^ note that in this solution, you'd be returning arrays.